funsor/joint.py

```diff
@@ -158,11 +158,7 @@ def moment_matching_contract_joint(red_op, bin_op, reduced_vars, discrete, gauss
 def eager_reduce_exp(op, arg, reduced_vars):
     # x.exp().reduce(ops.add) == x.reduce(ops.logaddexp).exp()
     log_result = arg.arg.reduce(ops.logaddexp, reduced_vars)
-    if log_result is not normalize.interpret(
-        Reduce, ops.logaddexp, arg.arg, reduced_vars
-    ):
```

Comment:
It's supposed to break a cycle in some of our rewrite rules that would otherwise lead to an infinite loop.

Reply:
I don't think it can be removed safely.

Reply:
One solution is to use `Tensor(torch.tensor(0))` instead of `Number(0)`, e.g.:

```py
scale += funsor.Constant(
    point.inputs,
    funsor.Tensor(ops.new_zeros(funsor.tensor.get_default_prototype(), ()))
)
```

Using Python numbers can be tricky with cons-hashing because they don't have any memory of past operations unlike `torch.Tensor` scalars:

```py
>>> a = 0
>>> b = a + 0
>>> a is b
True
>>> c = torch.tensor(0)
>>> d = c + 0
>>> c is d
False
```